### ISSUE TITLE
Proposal: steal.lock File Support

### ISSUE THREAD

**Issue by frank-dspeed** [stealjs/steal-tools#540]
steal-tools schould be able to create a lock file for a project like yarn does it after processing of a package.json.

This way we would have a point where steal-tools can access a complet list of all dependencys withour re sourcing the project.

think about it like a cache that enables commands like steal list

greetings.

usefull also for #538 and all other multi generation 

i think its usefull to know all requires and imports from all files in a single tree also will help with debuging and other features. Could be also added to steal main so that it can read the lockfile and prefer that if it exist over the config :)

**Comment by frank-dspeed** [comment]
http://stealjs.com/docs/steal-tools.BuildResult.html

**Comment by matthewp** [comment]
Yarn is a different type of tool and its lock file has a purpose; so every time you install you get the same versions of packages. I'm not sure what the purpose of a steal.lock would be, since steal doesn't install dependencies, it just uses the ones that you've installed.

> i think its usefull to know all requires and imports from all files in a single tree also will help with debuging and other features. 

I think a way to print out the dependency graph (and maybe export it to some format) does have some uses, I don't think it needs to be a lock file though (you're not really locking anything).

**Comment by frank-dspeed** [comment]
Yes thats what i whanted to discusse with you on gitter the lock file should be a dependency graph and should be a lock file to indicate that this graph is processed as the yarn.lock does it contains also the resolved promise infos so it needs no additional processing

goals:

From the graph aka lock we can make the individual builds realy realy easy with diffrent settings and also implament some nice functions from systemjs and overall be a sucessor.

**Comment by matthewp** [comment]
Let's back up for a second :-)  Instead of talking about solutions (a lock file or whatever else), let's talk about *problems*.

With this idea, what sort of problem are you hoping to solve?  You said:

> From the graph aka lock we can make the individual builds realy realy easy with diffrent settings and also implament some nice functions from systemjs and overall be a sucessor.

So what features are you looking to enable?

**Comment by frank-dspeed** [comment]
It will solve near all issus because it enables a complet new export and build process that will lead into more features and solve mainly and address:

#538 #512 #516 #446

**Comment by matthewp** [comment]
Yeah, I think we need to focus on problems for now. Solutions come later. I'm going to close because this doesn't have anything to do with #446 in particular, (or any of those other issues as far as I can tell).  If you are interested in a **feature** that you think lock files would solve, please create an issue about that feature. Then we can debate whether lock files are a good solution or not :)

**Comment by frank-dspeed** [comment]
Ah no Problem i keep it like always :) i will fork of the whole steal tool chain rename it and add the needed features only whanted to contribute back.

**Comment by frank-dspeed** [comment]
@matthewp i think the name makes you confused we put into that lock file all information about how this gets processed we examin it all via steal-tools and build with it a additional file for steal and steal tools where we put in our processing results.

Example Content of lock:

Complet dependency map from source
Detection of Code used in Source Like buffer maybe even frameworks all info we can get we get!
Realy all information we need gets into this
we could even cache load trees think about it also as cache

**Comment by matthewp** [comment]
@frank-dspeed can you provide an example of what this lock file will look like?  Like, just some text of what it will contain (whether yaml or json or whatever, i'm not sure):

```
foo-bar
  something-else:
    more-info
```

**Comment by frank-dspeed** [comment]
i think with the new production loader we should be able to do that this is not needed at present.

It is still usefull to do something like dry build and simply log some info to the console but a lock file is not needed. as all this could be done also in the config js parts of steal as far as i saw sorry for the trouble :)

I think all is in shape i need now only to produce a lot of donejs plugins and look what happens if i load them via steal and without it to check what needs improvements.
